include/deal.II/lac/trilinos_sparsity_pattern.h

```diff
@@ -1300,18 +1301,23 @@ namespace TrilinosWrappers
 
       ++accessor.a_index;
 
-      // If at end of line: do one
-      // step, then cycle until we
-      // find a row with a nonzero
-      // number of entries.
+      // If at end of line: do one step, then cycle until we find a row with a
+      // nonzero number of entries that is stored locally.
       if (accessor.a_index >= accessor.colnum_cache->size())
         {
           accessor.a_index = 0;
           ++accessor.a_row;
 
-          while ((accessor.a_row < accessor.sparsity_pattern->n_rows()) &&
-                 (accessor.sparsity_pattern->row_length(accessor.a_row) == 0))
-            ++accessor.a_row;
+          while (accessor.a_row < accessor.sparsity_pattern->n_rows())
+            {
+              const auto row_length =
+                accessor.sparsity_pattern->row_length(accessor.a_row);
+              if (row_length == 0 ||
+                  row_length == static_cast<SparsityPattern::size_type>(-1))
```

Comment:
does this indicate that we are at a line that is not locally stored? don't we have a less obscure way of checking this?

Reply:
No, here we can't simplify the logic. We are not storing the owned objects and `local_range` only report the first and last row stored. We use similar code in `add_entries` but access the `Epetra_FECrsGraph` object directly. We can't do that here since that object is `private`.

Reply:
I see. Yes, this is problematic when we use non-contiguous ranges.